internal/characters/tartaglia/abil.go

```diff
@@ -0,0 +1,411 @@
+package tartaglia
+
+import (
+	"fmt"
+
+	"github.com/genshinsim/gsim/pkg/core"
+)
+
+// Normal attack
+// Perform up to 6 consecutive shots with a bow.
+func (c *char) Attack(p map[string]int) (int, int) {
+	travel, ok := p["travel"]
+	if !ok {
+		travel = 20
+	}
+
+	f, a := c.ActionFrames(core.ActionAttack, p)
+
+	if c.Core.Status.Duration("tartagliamelee") > 0 {
+		return c.meleeAttack(f, a)
+	}
+
+	d := c.Snapshot(
+		fmt.Sprintf("Normal %v", c.NormalCounter),
+		core.AttackTagNormal,
+		core.ICDTagNormalAttack,
+		core.ICDGroupDefault,
+		core.StrikeTypePierce,
+		core.Physical,
+		25,
+		attack[c.NormalCounter][c.TalentLvlAttack()],
+	)
+
+	c.QueueDmg(&d, f+travel)
+
+	c.AdvanceNormalIndex()
+
+	return f, a
+}
+
+// random delayed numbers
+var meleeDelayOffset = [][]int{
+	{0},
+	{0},
+	{0},
+	{0},
+	{0},
+	{1, 0},
+}
+
+// Melee stance attack.
+// Perform up to 6 consecutive Hydro strikes.
+func (c *char) meleeAttack(f, a int) (int, int) {
+	for i, mult := range eAttack[c.NormalCounter] {
+		c.AddTask(func() {
+			d := c.Snapshot(
+				fmt.Sprintf("Normal %v", c.NormalCounter),
+				core.AttackTagNormal,
+				core.ICDTagNormalAttack,
+				core.ICDGroupDefault,
+				core.StrikeTypeSlash,
+				core.Hydro,
+				25,
+				mult[c.TalentLvlSkill()],
+			)
+			d.OnHitCallback = c.rtSlashCallback
```

Comment:
This is going to seem somewhat cursed, but I realized in game that riptide applications from melee attacks always apply first, so if you apply riptide with a melee crit, then the riptide damage procs immediately. Should check that happens here.

Note that it doesn't work this way for CAs - if you apply riptide from CAs, then Riptide flash does not immediately proc.

Reply:
Cursed indeed. I didn't find a right way to get it done with current riptide slash callback. So I will let how it is rn